### PR TITLE
Unify and improve const-prop lints

### LINKED ISSUE #69020: Duplicate errors for overflowing division / remainder
When there is an overflowing division or remainder operation that const_prop can detect, the error is duplicated *in release mode only*:
```rust
#![deny(const_err)]

use std::i32;

fn main() {
    let _ = i32::MIN / -1;
}
```
shows
```
error: attempt to divide with overflow
 --> src/main.rs:6:13
  |
6 |     let _ = i32::MIN / -1;
  |             ^^^^^^^^^^^^^
  |
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![deny(const_err)]
  |         ^^^^^^^^^

error: this expression will panic at runtime
 --> src/main.rs:6:13
  |
6 |     let _ = i32::MIN / -1;
  |             ^^^^^^^^^^^^^ attempt to divide with overflow
```
This is because `!overflow_check` is `true` in release mode but divison and remainder *still* get overflow checks.

([This test case](https://github.com/rust-lang/rust/blob/4d1241f5158ffd66730e094d8f199ed654ed52ae/src/test/ui/issues/issue-8460-const.rs#L12) already covers the problem, so a fix does not need a new test case, it just needs to adjust the existing test case to no longer expect two `ERROR` per line.)

Cc @oli-obk @wesleywiser

### PR DESCRIPTION
**Pull request by RalfJung**
Add a single helper method for all lints emitted by const-prop, and make that lint different from the CTFE `const_err` lint. Also consistently check overflow on *arithmetic*, not on the assertion, to make behavior the same for debug and release builds.

See [this summary comment](https://github.com/rust-lang/rust/pull/69185#issuecomment-587924754) for details and the latest status.

In terms of lint formatting, I went for what seems to be the better style: have a general message above the code, and then a specific message at the span:
```
error: this arithmetic operation will overflow
  --> $DIR/const-err2.rs:21:18
   |
LL |     let a_i128 = -std::i128::MIN;
   |                  ^^^^^^^^^^^^^^^ attempt to negate with overflow
```
We could also just have the specific message above and no text at the span if that is preferred.

I also converted some of the existing tests to use compiletest revisions, so that the same test can check a bunch of different compile flags.

Fixes https://github.com/rust-lang/rust/issues/69020.
Helps with https://github.com/rust-lang/rust/issues/69021: debug/release are now consistent, but the assoc-const test in that issue still fails (there is a FIXME in the PR for this). The reason seems to be that const-prop notices the assoc const in `T::N << 42` and does not even bother calling `const_prop` on that operation.
Has no effect on https://github.com/rust-lang/rust/issues/61821; the duplication there has entirely different reasons.

### CODE REVIEW

File: src/librustc_session/lint/builtin.rs [hunk 1]
```diff
@@ -46,6 +46,18 @@ declare_lint! {
     "shift exceeds the type's number of bits"
 }
 
+declare_lint! {
+    pub OVERFLOW,
```
Comment (RalfJung) [inline src/librustc_session/lint/builtin.rs]: My inclination would be to merge this with `EXCEEDING_BITSHIFTS` into one lint, but I am not sure what the procedure is for changing/removing existing lint names.

Also, the new lint names are maybe not great, suggestions are welcome :)

File: src/librustc_session/lint/builtin.rs [hunk 2]
```diff
@@ -46,6 +46,18 @@ declare_lint! {
...
+    Deny,
```
Comment (RalfJung) [inline src/librustc_session/lint/builtin.rs]: I'd lean towards warn-by-default, but `EXCEEDING_BITSHIFTS` was deny-by-default so I stuck with that.

File: src/test/ui/consts/const-err2.rs [hunk 3]
```diff
@@ -2,33 +2,34 @@
 // optimized compilation and unoptimized compilation and thus would
 // lead to different lints being emitted
 
+// revisions: default noopt opt opt_with_overflow_checks
```
Comment (RalfJung) [inline src/test/ui/consts/const-err2.rs]: I may have gone overboard by having 4 revisions -- specifically, `default` (no extra flags) is likely a duplicate of one of the others, but I am not sure which. We already have *so many* tests, so I am somewhat inclined to remove `default` again... any opinions?

File: src/test/ui/consts/const-eval/promoted_errors.rs [hunk 4]
```diff
@@ -1,23 +1,28 @@
...
 // build-pass
 // ignore-pass (emit codegen-time warnings and verify that they are indeed warnings and not errors)
-// compile-flags: -O
 
-#![warn(const_err)]
+#![warn(const_err, overflow, panic)]
 
 fn main() {
     println!("{}", 0u32 - 1);
+    //[opt_with_overflow_checks,noopt]~^ WARN [overflow]
```
Comment (RalfJung) [inline src/test/ui/consts/const-eval/promoted_errors.rs]: This makes it look like `default` is the same as `opt`, which I find surprising.

File: src/test/ui/issues/issue-8460-const.rs [hunk 5]
```diff
@@ -8,63 +12,51 @@ use std::thread;
...
+    //~^ ERROR operation will panic
     assert!(thread::spawn(move|| { 1i64 / 0; }).join().is_err());
-    //~^ ERROR attempt to divide by zero
+    //~^ ERROR operation will panic
     assert!(thread::spawn(move|| { 1i128 / 0; }).join().is_err());
-    //~^ ERROR attempt to divide by zero
+    //~^ ERROR operation will panic
     assert!(thread::spawn(move|| { isize::MIN % -1; }).join().is_err());
-    //~^ ERROR attempt to calculate the remainder with overflow
-    //~| ERROR this expression will panic at runtime
```
Comment (RalfJung) [inline src/test/ui/issues/issue-8460-const.rs]: Yay, duplicates are gone :)

File: src/test/compile-fail/consts/const-err3.rs [hunk 6]
```diff
@@ -7,13 +7,13 @@ fn black_box<T>(_: T) {
 
 fn main() {
     let b = 200u8 + 200u8 + 200u8;
-    //~^ ERROR const_err
```
Comment (RalfJung) [inline src/test/compile-fail/consts/const-err3.rs]: I am not entirely sure of the purpose of this test, it looks entirely like a duplicate of `const-err2.rs`...

Comment (oli-obk) [reply src/test/compile-fail/consts/const-err3.rs]: Iirc I duplicates this because there was some abort-if-errors call that caused some errors not to get emitted

File: src/librustc_session/lint/builtin.rs (hunk 2 above)
Comment (Centril) [reply src/librustc_session/lint/builtin.rs]: I think a refactoring PR isn't the place to change the severity of the lint.

File: src/librustc_mir/transform/const_prop.rs [hunk 7]
```diff
@@ -501,19 +500,20 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
         }
     }
 
-    fn report_panic_as_lint(&self, source_info: SourceInfo, panic: AssertKind<u64>) -> Option<()> {
-        // Somewhat convoluted way to re-use the CTFE error reporting code.
+    fn report_assert_as_lint(
+        &self,
+        lint: &'static lint::Lint,
+        source_info: SourceInfo,
+        message: &str,
```
Comment (Centril) [inline src/librustc_mir/transform/const_prop.rs]:
```suggestion
        message: &'static str,
```
The `struct_span_lint_hir` infra with `.build(message)` was intentionally setup in a way to discourage allocation of messages in case the lint doesn't fire. In all cases you have a `&'static str`, but `: &str` could invite doing `&format!(...)`.

File: src/test/ui/consts/const-err2.rs (hunk 3 above)
Comment (RalfJung) [reply src/test/ui/consts/const-err2.rs]: Sure but there's such a thing as too much of any good thing. ;)

File: src/test/compile-fail/consts/const-err3.rs (hunk 6 above)
Comment (RalfJung) [reply src/test/compile-fail/consts/const-err3.rs]: So, can I remove it?

File: src/librustc_mir/transform/const_prop.rs (hunk 7 above)
Comment (RalfJung) [reply src/librustc_mir/transform/const_prop.rs]: I don't feel like the cost of an allocation is relevant here but whatever.

File: src/librustc_session/lint/builtin.rs (hunk 2 above)
Comment (RalfJung) [reply src/librustc_session/lint/builtin.rs]: This is not a refactoring PR. This is the part that I did not do in https://github.com/rust-lang/rust/pull/68969 because *that* was a refactoring PR.
For example, this PR already moves a bunch of warnings from `const_err` to other lints.

File: src/librustc_session/lint/builtin.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_session/lint/builtin.rs]: Well, having a separate lint for bitshifts seems silly. I mean, if that's the way to go we should have separate lints for negation, addition, subtraction, ...

File: src/librustc_session/lint/builtin.rs (hunk 2 above)
Comment (RalfJung) [reply src/librustc_session/lint/builtin.rs]: That said, I am fine with keeping the severity as-is. I don't have a strong opinion. This just feels more like "warn" material to me than "deny". If @rust-lang/wg-diagnostics agrees, someone can file a follow-up to change the default.

File: src/test/compile-fail/consts/const-err3.rs (hunk 6 above)
Comment (oli-obk) [reply src/test/compile-fail/consts/const-err3.rs]: As long as both tests have the same diagnostics emitted at this time, yes

File: src/librustc_session/lint/builtin.rs (hunk 1 above)
Comment (oli-obk) [reply src/librustc_session/lint/builtin.rs]: I agree, let's do it in a separate PR though and ping some ppl on it

Comment (Centril) [reply src/librustc_session/lint/builtin.rs]: Seems like a design question that ought to be bumped to the language team?

Comment (RalfJung) [reply src/librustc_session/lint/builtin.rs]: Is the lang team involved with the details of lint naming? I would have said this is compiler team territory but what do I know.^^

File: src/librustc_session/lint/builtin.rs [hunk 8]
```diff
@@ -46,6 +46,18 @@ declare_lint! {
...
+    "arithmetic operation overflows"
+}
+
+declare_lint! {
+    pub PANIC,
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/librustc_session/lint/builtin.rs]: That sounds like `panic!("...")`, which is indeed unconditional but doesn't fit.

Under the hood this is an assertion failure, but that might also be confusing because it sounds like `assert!`.

File: src/librustc_session/lint/builtin.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_session/lint/builtin.rs]: > How about overflowing_expressions

An alternative is `arithmetic_overflow`.

Comment (RalfJung) [reply src/librustc_session/lint/builtin.rs]: OTOH, if we want to extend this to include shifts later... do they count as "arithmetic"?

Comment (Centril) [reply src/librustc_session/lint/builtin.rs]: The lang team probably doesn't care about naming per se, but granularity and lint levels, probably yes.

**Comment by RalfJung** [comment]
Man, is there any part of the test suite that does *not* check these lints?^^

File: src/librustc_session/lint/builtin.rs (hunk 1 above)
Comment (Centril) [reply src/librustc_session/lint/builtin.rs]: Although if we can reach a reasonable conclusion here that has some consensus re. grouping then I think we could also keep things here and not "escalate".

File: src/librustc_session/lint/builtin.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (oli-obk) [reply src/librustc_session/lint/builtin.rs]: So it's an "always failing dynamic check". No idea how to name this nicely

File: src/librustc_session/lint/builtin.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_session/lint/builtin.rs]: Well, one thing we have to do in this PR is pick a name for the new lint. Proposals so far, and some new ones:
* overflow
* overflowing_expressions
* arithmetic_overflow
* integer_overflow

Whether or not this also covers overflowing bitshifts is orthogonal insofar as that doesn't impact the choice of name.

(Also I made the choice of making this lint separate from the one that warns about panics. They were previously all together, but the thing is that overflow doesn't always panic, so it seems to make sense to separate that.)

File: src/librustc_session/lint/builtin.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/librustc_session/lint/builtin.rs]: Despite my reservations, `unconditional_panic` is my favorite so far.

**Comment by RalfJung** [comment]
I now picked `arithmetic_overflow` and `unconditional_panic` as lint names. I also decided, because this PR already renames lints, it is better to do all renaming at once, so `exceeding_bitshifts` is now also covered by `arithmetic_overflow`. However, the lint level remains the same: deny-by-default.

---

So, in case you want to FCP or otherwise want a summary: what this PR does is to take *some of* the lints that were so far reported as `const_err`, and move them to a different category. `const_err` ought to be for hard errors during const-eval, but we were also using it for when we determined that some run-time code was likely buggy (e.g. because it has overflowing arithmetic or an operation that will always panic, such as div-by-0). This analysis is performed as a side-effect of const-prop (*not* cons-eval).
Those cases (and only those) are now reported as `arithmetic_overflow` and `unconditional_panic`, respectively. It also turns out that const-prop already used a dedicated lint for one particular kind of overflow, namely `exceeding_bitshifts`. So this PR also changes those lints to be emitted as `arithmetic_overflow`, because it seems odd to have one lint per operation and now that we are touching the lints anyway we might as well do something that makes sense.

In other words, no new lints get emitted by this PR, but the following changes:

* Parts of `const_err` become `unconditional_panic`.
* Other parts of `const_err` become `arithmetic_overflow`.
* `exceeding_bitshifts` becomes `arithmetic_overflow`.

With this done, one rarely wants to allow `const_err` (this lint is now only fired for constants that cannot be used because their evaluation failed [including promoteds]); the only reason that is a lint and not a hard error is backwards-compatibility. The other lints however are more like our regular diagnostics. I think we should make them warn-by-default instead of deny-by-default, but for now this PR leaves the default levels unchanged (but I am happy to adjust in this PR or a follow-up).

**Comment by Centril** [comment]
I'm happy with the new lint names. I don't think FCP for renaming necessary. I would like to leave the lint severities as-is though. The enumerated lints seem like a case of correctness class lints that highlight actual bugs, so I think stopping compilation (as achieved with deny-by-default) is warranted.

**Comment by RalfJung** [comment]
All right. @oli-obk the PR is awaiting your review, then. :)

**Comment by RalfJung** [comment]
I marked this for relnotes, given the user-visible changes here. I hope that is appropriate.
